allennlp/training/callbacks/callbacks.py

```diff
@@ -0,0 +1,512 @@
+# These should probably all live in separate files
+
+from typing import Set, Dict, List, Tuple, Optional
+import copy
+import datetime
+import logging
+import math
+import os
+import time
+import traceback
+
+import torch
+
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.params import Params
+from allennlp.common.tqdm import Tqdm
+from allennlp.common.util import lazy_groups_of, dump_metrics, gpu_memory_mb, peak_memory_mb
+from allennlp.models import Model
+from allennlp.training.optimizers import Optimizer
+from allennlp.training import util as training_util
+from allennlp.training import trainer2  # pylint: disable=unused-import
+from allennlp.training.checkpointer import Checkpointer
+from allennlp.training.callbacks import Callback, Events
+from allennlp.training.learning_rate_schedulers import LearningRateScheduler
+from allennlp.training.metric_tracker import MetricTracker
+from allennlp.training.momentum_schedulers import MomentumScheduler
+from allennlp.training.moving_average import MovingAverage
+from allennlp.training.tensorboard_writer import TensorboardWriter
+
+logger = logging.getLogger(__name__)
+
+
+@Callback.register("log_to_tensorboard")
+class LogToTensorboard(Callback['trainer2.Trainer']):
+    """
+    Callback that handles all Tensorboard logging.
+    """
+    def __init__(self,
+                 tensorboard: TensorboardWriter,
+                 log_batch_size_period: int = None) -> None:
+        self.log_batch_size_period = log_batch_size_period
+        self.tensorboard = tensorboard
+
+        self.cumulative_batch_size = 0
+
+        # For logging histograms
+        self.histogram_parameters: Set[str] = set()
+        self.param_updates: Dict[str, torch.Tensor] = {}
+
+    def __call__(self, event: str, trainer: 'trainer2.Trainer') -> None:
+        if event == Events.TRAINING_START:
```

Comment:
I much prefer @epwalsh's idea of using decorators on methods instead of this giant `__call__` method.  At the very least you should split out each block inside this `if/else` into its own method.  And once you do that, `__call__` becomes unnecessary if we can just decorate those methods directly - this simplifies the implementation of every callback, because the handler has logic to do the routing.